### ISSUE TITLE
README.md

### ISSUE THREAD

**Issue by starbasessd** [goldstar611/chirp-appimage#1]
I am confused. I am trying to follow the README, but the path is very broken. I don't understand how much of the tutorial is needed, where you ran into issues, etc. If you look at the [Installation instructions](https://github.com/ccrisan/motioneye/wiki/Installation) it would be helpful to break out exactly what steps & the commands you used to get a working install, especially for a novice like me.

**Comment by goldstar611** [comment]
Hi starbasessd,

You should be good to go after running the first block:
```
sudo apt install git python3-pip python3-setuptools patchelf desktop-file-utils libgdk-pixbuf2.0-dev fakeroot strace fuse

# Install appimagetool AppImage
sudo wget https://github.com/AppImage/AppImageKit/releases/download/continuous/appimagetool-x86_64.AppImage -O /usr/local/bin/appimagetool
sudo chmod +x /usr/local/bin/appimagetool
```

If you see any errors you can paste them here and I can instruct you further.

After the first block is successfully completed, you can move on to the last block
```
git clone https://github.com/goldstar611/chirp-appimage
cd chirp-appimage/
appimage-builder --recipe recipe.yml
```

**Comment by goldstar611** [comment]
I'm only familiar with package management in debian/ubuntu so I can't provide commands or help in the first block for other distributions like Fedora, Red Hat, Arch etc..

**Comment by starbasessd** [comment]
No problem. Will test in a bit, working on issue in motionEye for a user right now.

**Comment by starbasessd** [comment]
RPM (Fedora, RHEL, CentOS etc) and Arch are fairly easy.  BTDT, got a few t-shirts...

**Comment by goldstar611** [comment]
I've added `python2.7-minimal` to the first block. I've also added a `build.sh` script that will generate binaries for 64bit intel, 64bit arm and 32bit arm. Run `git pull` to refresh the files in the repository.

**Comment by starbasessd** [comment]
pi@deb11-flatpak:~/chirp-appimage$ appimage-builder --recipe recipe.yml
Traceback (most recent call last):
  File "/usr/local/bin/appimage-builder", line 8, in <module>
    sys.exit(__main__())
  File "/usr/local/lib/python3.9/dist-packages/appimagebuilder/__main__.py", line 48, in __main__
    raw_recipe_data = recipe_loader.load(args.recipe)
  File "/usr/local/lib/python3.9/dist-packages/appimagebuilder/recipe/loader.py", line 76, in load
    return yaml.load(recipe_file, Loader=self._loader)
  File "/usr/lib/python3/dist-packages/yaml/__init__.py", line 114, in load
    return loader.get_single_data()
  File "/usr/lib/python3/dist-packages/yaml/constructor.py", line 51, in get_single_data
    return self.construct_document(node)
  File "/usr/lib/python3/dist-packages/yaml/constructor.py", line 60, in construct_document
    for dummy in generator:
  File "/usr/lib/python3/dist-packages/yaml/constructor.py", line 413, in construct_yaml_map
    value = self.construct_mapping(node)
  File "/usr/lib/python3/dist-packages/yaml/constructor.py", line 218, in construct_mapping
    return super().construct_mapping(node, deep=deep)
  File "/usr/lib/python3/dist-packages/yaml/constructor.py", line 143, in construct_mapping
    value = self.construct_object(value_node, deep=deep)
  File "/usr/lib/python3/dist-packages/yaml/constructor.py", line 100, in construct_object
    data = constructor(self, node)
  File "/usr/local/lib/python3.9/dist-packages/appimagebuilder/recipe/loader.py", line 64, in constructor_env_variables
    raise RecipeError(
appimagebuilder.recipe.errors.RecipeError: Unable to resolve environment variable: TARGET_ARCH_APPIMAGE

**Comment by starbasessd** [comment]
pi@deb11-flatpak:~/chirp-appimage$ ./build.sh
+ rm -f 'Chirp-*-*.AppImage'
+ git clone --depth=1 https://github.com/kk7ds/chirp
+ true
fatal: destination path 'chirp' already exists and is not an empty directory.
+ cd chirp
+ git pull
hint: Pulling without specifying how to reconcile divergent branches is
hint: discouraged. You can squelch this message by running one of the following
hint: commands sometime before your next pull:
hint:
hint:   git config pull.rebase false  # merge (the default strategy)
hint:   git config pull.rebase true   # rebase
hint:   git config pull.ff only       # fast-forward only
hint:
hint: You can replace "git config" with "git config --global" to set a default
hint: preference for all repositories. You can also pass --rebase, --no-rebase,
hint: or --ff-only on the command line to override the configured default per
hint: invocation.
Already up to date.
++ git log -1 --date=format:%Y%m%d --format=%ad
+ LATEST_COMMIT_DATE=20220227
++ git rev-parse --short HEAD
+ LATEST_COMMIT_SHORT=799d07a
+ export CHIRP_VERSION=daily-20220227+799d07a
+ CHIRP_VERSION=daily-20220227+799d07a
+ cd ..
+ export 'SOURCE_LINE_1=deb [arch=amd64] http://archive.ubuntu.com/ubuntu/ bionic main restricted universe multiverse'
+ SOURCE_LINE_1='deb [arch=amd64] http://archive.ubuntu.com/ubuntu/ bionic main restricted universe multiverse'
+ export 'SOURCE_LINE_2=deb [arch=amd64] http://archive.ubuntu.com/ubuntu/ xenial main restricted universe multiverse'
+ SOURCE_LINE_2='deb [arch=amd64] http://archive.ubuntu.com/ubuntu/ xenial main restricted universe multiverse'
+ export TARGET_ARCH_APT=amd64
+ TARGET_ARCH_APT=amd64
+ export TARGET_ARCH_APPIMAGE=x86_64
+ TARGET_ARCH_APPIMAGE=x86_64
+ appimage-builder --recipe recipe.yml
INFO:main:Running main script
rm -rf AppDir  | true
mkdir -p AppDir/usr/share/gtk-2.0/
echo 'include "/usr/share/themes/Adwaita/gtk-2.0/gtkrc"' > AppDir/usr/share/gtk-2.0/gtkrc
cd chirp/
# Update in-app version
sed -i "s/CHIRP_VERSION = \".*\"/CHIRP_VERSION = \"${CHIRP_VERSION}\"/" chirp/__init__.py
# Install chirp into the AppDir
python2.7 setup.py install --prefix=/usr --root=../AppDir
Traceback (most recent call last):
  File "setup.py", line 6, in <module>
    from chirp.drivers import *
  File "/home/pi/chirp-appimage/chirp/chirp/drivers/idrp.py", line 16, in <module>
    import serial
ImportError: No module named serial
Traceback (most recent call last):
  File "/usr/local/bin/appimage-builder", line 8, in <module>
    sys.exit(__main__())
  File "/usr/local/lib/python3.9/dist-packages/appimagebuilder/__main__.py", line 58, in __main__
    invoker.execute(commands)
  File "/usr/local/lib/python3.9/dist-packages/appimagebuilder/invoker.py", line 41, in execute
    command()
  File "/usr/local/lib/python3.9/dist-packages/appimagebuilder/commands/run_shell_script.py", line 64, in __call__
    raise RuntimeError("Script exited with code: %s" % _proc.returncode)
RuntimeError: Script exited with code: 1
pi@deb11-flatpak:~/chirp-appimage$

**Comment by goldstar611** [comment]
Ok build script is updated again so that the host doesn't need pyserial or other python packages. Run `git pull` again and restart the build.sh script

**Comment by starbasessd** [comment]
Something is hard-coded to Ubuntu, I'm running Debian...
Writing ../AppDir/usr/lib/python2.7/site-packages/chirp-daily_20220227_799d07a-py2.7.egg-info
# https://github.com/AppImageCrafters/appimage-builder/pull/186
mkdir ${APPDIR}/usr/share/pixmaps/255x104/
cp share/chirp.png ${APPDIR}/usr/share/pixmaps/255x104/chirp.png
INFO:main:Running apt deploy
INFO:apt:apt-get update
Get:1 http://archive.ubuntu.com/ubuntu bionic InRelease [242 kB]
Get:2 http://archive.ubuntu.com/ubuntu xenial InRelease [247 kB]
Err:1 http://archive.ubuntu.com/ubuntu bionic InRelease
  The following signatures couldn't be verified because the public key is not available: NO_PUBKEY 3B4FE6ACC0B21F32
Err:2 http://archive.ubuntu.com/ubuntu xenial InRelease
  The following signatures couldn't be verified because the public key is not available: NO_PUBKEY 40976EAF437D05B5 NO_PUBKEY 3B4FE6ACC0B21F32
Reading package lists... Done
W: GPG error: http://archive.ubuntu.com/ubuntu bionic InRelease: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY 3B4FE6ACC0B21F32
E: The repository 'http://archive.ubuntu.com/ubuntu bionic InRelease' is not signed.
N: Updating from such a repository can't be done securely, and is therefore disabled by default.
N: See apt-secure(8) manpage for repository creation and user configuration details.
W: GPG error: http://archive.ubuntu.com/ubuntu xenial InRelease: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY 40976EAF437D05B5 NO_PUBKEY 3B4FE6ACC0B21F32
E: The repository 'http://archive.ubuntu.com/ubuntu xenial InRelease' is not signed.
N: Updating from such a repository can't be done securely, and is therefore disabled by default.
N: See apt-secure(8) manpage for repository creation and user configuration details.
ERROR:root:"apt-get update" execution failed
Traceback (most recent call last):
  File "/usr/local/bin/appimage-builder", line 8, in <module>
    sys.exit(__main__())
  File "/usr/local/lib/python3.9/dist-packages/appimagebuilder/__main__.py", line 58, in __main__
    invoker.execute(commands)
  File "/usr/local/lib/python3.9/dist-packages/appimagebuilder/invoker.py", line 41, in execute
    command()
  File "/usr/local/lib/python3.9/dist-packages/appimagebuilder/commands/apt_deploy.py", line 46, in __call__
    deployed_packages = apt_deploy.deploy(
  File "/usr/local/lib/python3.9/dist-packages/appimagebuilder/modules/deploy/apt/deploy.py", line 41, in deploy
    self._prepare_apt_venv()
  File "/usr/local/lib/python3.9/dist-packages/appimagebuilder/modules/deploy/apt/deploy.py", line 50, in _prepare_apt_venv
    self.apt_venv.update()
  File "/usr/local/lib/python3.9/dist-packages/appimagebuilder/modules/deploy/apt/venv.py", line 160, in update
    shell.assert_successful_result(_proc)
  File "/usr/local/lib/python3.9/dist-packages/appimagebuilder/utils/shell.py", line 36, in assert_successful_result
    raise RuntimeError(
RuntimeError: "apt-get update" execution failed with code 100

**Comment by goldstar611** [comment]
Yes, the script is using the Ubuntu package repository since all of GTK and other packages have already been compiled. We just download and extract them essentially.

I've added in the 2 keys mentioned in the error. `git pull` and try it again. Youre feedback is very valuable. Thanks!

**Comment by starbasessd** [comment]
perl and sh can't be found
example WARNINGs:
WARNING:root:Required interpreter 'perl' could not be found in the AppDir while processing /home/pi/chirp-appimage/AppDir/usr/bin/edit
WARNING:root:Required interpreter 'sh' could not be found in the AppDir while processing /home/pi/chirp-appimage/AppDir/usr/share/python/runtime.d/public_modules.rtinstall
INFO:root:AppImage created successfully
Where is the appimage executable supposed to be?

**Comment by starbasessd** [comment]
BTW sh is at /usr/bin/sh
perl is at /usr/bin/perl and perl --version reports Perl 5 version 32 subversion 1

**Comment by goldstar611** [comment]
We dont use sh or pearl in our AppImage so those can be safely ignored :)

**Comment by goldstar611** [comment]
> Where is the appimage executable supposed to be?
You should see `Chirp-daily-20220227+799d07a-aarch64.AppImage` and `Chirp-daily-20220227+799d07a-armhf.AppImage` and `Chirp-daily-20220227+799d07a-x86_64.AppImage` in the current directory. run an `ls` command or find them in your file browser. You did it!

**Comment by starbasessd** [comment]
IIRC doesn't chirp use them?

**Comment by goldstar611** [comment]
I don't think so. Chirp was written in python (not perl). There are no shell scripts that run only python scripts.

**Comment by starbasessd** [comment]
BTW, gives me an "Update as soon as possible (20220219 latest version) but seems to be up and running.

**Comment by goldstar611** [comment]
> BTW, gives me an "Update as soon as possible (20220219 latest version) but seems to be up and running.

I need to work with Dan on how he does his update logic. Let me know if you can read from a radio (if you have one near by). Thank you so much for your help and running it in a clean environment. That helped tremendously.

**Comment by starbasessd** [comment]
It's running in a VM, will take me a bit to get the USB registered to it.
I'll check the armhf in a bit, too...

**Comment by starbasessd** [comment]
Boom, shaka laka... LOVE virtualbox since they fixed USB... Just read a Baofeng UV5RA into the new Chirp. Excellent. May be later tonight before I can test the armhf on a Pi, though. Grandsons are home...

**Comment by goldstar611** [comment]
Great news! Enjoy your time with the little guys! Many thanks again.

**Comment by starbasessd** [comment]
If you work with me, I'll bend over backwards to help you out... This gives me other ideas, too...

**Comment by starbasessd** [comment]
What does the aarch64 image cover?

**Comment by goldstar611** [comment]
> What does the aarch64 image cover?

Any 64-bit ARM device running a 64-bit OS. For example, my raspberry pi 4 has a 64-bit ARM chip but a 32-bit OS, so I have to run the armhf binary. You can find which OS you have by running `uname -a`

```
pi@raspberrypi4:~ $ uname -a
Linux raspberrypi4 5.10.63-v7l+ #1459 SMP Wed Oct 6 16:41:57 BST 2021 armv7l GNU/Linux
```

That right most text `armv7l GNU/Linux`, specifically armv7 means I'm on a 32-bit OS. If that would say `aarch64` or `arm64` then I could use the aarch64 bit binary.

**Comment by starbasessd** [comment]
I know about the differences (deal with them all the time...) just needed confirmation... Was more concerned about ARM vs M1 ARM... since theoretically the new Apple chipset is a bit different... less than 100MB for all three combined, Jason over at Build-A-Pi might be interested, too... He's having issues with Chirp, too...

**Comment by starbasessd** [comment]
What would be needed to add a 32 bit image for x86? (older laptops in the field used for programming?)

**Comment by goldstar611** [comment]
Oh the M1 ARM. Sadly I don't think AppImages are compatible with MacOS.

> What would be needed to add a 32 bit image for x86? (older laptops in the field used for programming?)

Just a few more lines of code in build.sh to set up the correct apt repository. Want me to add that?

**Comment by starbasessd** [comment]
Sure, I am currently switching over to RPis myself (trashed my last old laptop last Field Day) but if you can extend the life for others, it would probably be appreciated.
I can test her with a VM, still.

**Comment by goldstar611** [comment]
> but if you can extend the life for others, it would probably be appreciated.

build script updated, `git pull` to get the changes

> I can test here with a VM, still.

That would be super helpful!

**Comment by starbasessd** [comment]
Building...

**Comment by starbasessd** [comment]
Working fine, thanks...

**Comment by goldstar611** [comment]
I don't know @km4ack but if he is willing to give this chirp AppImage a try on a raspberry pi that would be awesome!

Sorry to bring you in on such a long thread but @starbasessd and I have been working to get chirp into an "AppImage" and he mentioned you were having some issues getting chirp running on a pi. Would you be interested in testing?

**Comment by km4ack** [comment]
@goldstar611 Someone say raspberry pi? Yep, I'm in. I'll read up on what you guys have going and give it a shot. I may have some time tomorrow afternoon. Thanks for giving me a heads up on this project!

**Comment by goldstar611** [comment]
Yay! You don't have to read the whole thread or even compile the AppImage yourself unless  you would like to learn more about the build process.

I have builds posted at https://github.com/goldstar611/chirp-appimage/releases/tag/02Mar2022 No installation needed. 

You just need to grab the right binary for your CPU+OS type. 
- raspberry pi 4 with 32bit os -> armhf
- raspberry pi 4 with 64bit os -> aarch64
- recent Linux desktop computer ->  x86_64

**Comment by goldstar611** [comment]
> I am terrible about not reading/following directions so I downloaded the armhf version instead of the aarch64 version onto a Pi 4 running Pi OS 64bit. 

I would have done the same thing!

> Chirp started without issue. Verified it would read from a Baofeng. All seems to be working great even though I picked the wrong version. 

That's great news. Which Baofeng model did you test with? I'd like to add it to my list

> I did get a notification about a new release being available. I am flashing a new SD card now and will test the correct version for 64 bit later today.

This is an artifact of me not knowing how Dan does his release numbers. Once I hand over the scripts to Dan this should be sorted out.

> Since the Chirp build changes almost daily, I assume I will need to build everything from source to have the latest version installed?

My plan is to hand off the work I've done to Dan and that he would either replace the current flatpak files or add them to the list of downloads at https://trac.chirp.danplanet.com/chirp_daily/LATEST/

**Comment by starbasessd** [comment]
Computers tested: 
X86-64 & X86-i686 Debian on Ryzen-7 processor VirtualBox VMs (2 core i686, 4 core x86-64)  4GB RAM, 50GB HD, Debian 9 x86-64 & i686, 10 x86-64 & i686, 11 x86-64. All default desktops.
Pi3B+ and Pi4B-4GB, RaspberryPiOS 32bit and 64bit, all fully updated. All booting to 32GB USB Drive.
All systems had to 'sudo usermod -aG dialout [user]' to enable access to /dev/ttyUSB0
Using the appropriate AppImage all ran correctly.
Used Baofeng UV5RA and Juentai JT6800mini with appropriate cable in all test OSs
All were able to download and upload without issue.
- - - - -
Tried to run in WSL  Debian on Win10 (Why? Why not!) but failed (also use MobaXterm (lets me run graphical apps as needed):
- - - - -
pi@Lenovo1:~$ ./Chirp-daily-20220227+799d07a-x86_64.AppImage
Traceback (most recent call last):
  File "/tmp/.mount_Chirp-tlgEk3//usr/bin/chirpw", line 131, in <module>
    from chirp.ui import mainapp
  File "/tmp/.mount_Chirp-tlgEk3/usr/lib/python2.7/site-packages/chirp/ui/mainapp.py", line 26, in <module>
    import gtk
  File "/tmp/.mount_Chirp-tlgEk3/usr/lib/python2.7/dist-packages/gtk-2.0/gtk/__init__.py", line 40, in <module>
    from gtk import _gtk
ImportError: libxcb-shm.so.0: cannot open shared object file: No such file or directory
pi@Lenovo1:~
- - - - -
Can't have everything, I guess. 
If you want anything else tested, let me know...

**Comment by km4ack** [comment]
I am using a Baofeng UV5r for testing. On a clean install of Pi OS 64 bit I downloaded the aarch64 this time. When trying to run I got an error (dlopen(): error loading libfuse.so.2). Installed libfuse2 and then the appimage opened as expected. Again I was able to read from the radio.

**Comment by km4ack** [comment]
On the Pi OS 64 bit system, I attempted building the app image. I ran these commands
```
sudo apt install git python2.7-minimal python3-pip python3-setuptools patchelf desktop-file-utils libgdk-pixbuf2.0-dev fakeroot strace fuse

sudo wget https://github.com/AppImage/AppImageKit/releases/download/continuous/appimagetool-x86_64.AppImage -O /usr/local/bin/appimagetool

sudo chmod +x /usr/local/bin/appimagetool

git clone https://github.com/goldstar611/chirp-appimage
cd chirp-appimage/
chmod +x ./build.sh
./build.sh
```
and it exited with error:
./build.sh: line 26: appimage-builder: command not found

**Comment by starbasessd** [comment]
As soon as the B-A-P gets done doing its thing, I'll test.

**Comment by goldstar611** [comment]
> and it exited with error:
> ./build.sh: line 26: appimage-builder: command not found

This is totally my fault, I removed the line on how to install appimage-builder in the readme. It's added back

```
sudo -H pip3 install appimage-builder
```

**Comment by starbasessd** [comment]
no *,AppImage files created,,,

**Comment by starbasessd** [comment]
I think I figured it out: 
> sudo wget https://github.com/AppImage/AppImageKit/releases/download/continuous/appimagetool-x86_64.AppImage -O /usr/local/bin/appimagetool

needs to be:
sudo wget https://github.com/AppImage/AppImageKit/releases/download/continuous/appimagetool-aarch64.AppImage -O /usr/local/bin/appimagetool

for PiOS-64bit (need to either note it in the instructions to wget the correct version or do something to 'fill it in' automagically...

**Comment by goldstar611** [comment]
yes exactly. let me see if I can reliably hook into `uname` for this step

**Comment by goldstar611** [comment]
What does `uname -m` give you on your 64-bit pi?

**Comment by starbasessd** [comment]
aarch64

**Comment by starbasessd** [comment]
Debian 11-64 reports x86_64
OS: Debian GNU/Linux 10 (buster) i686 reports:pi@rpi-i386:~ $ uname -m
i686

**Comment by goldstar611** [comment]
> Debian 11-64 reports x86_64 OS: Debian GNU/Linux 10 (buster) i686 reports:pi@rpi-i386:~ $ uname -m i686

That's great, the only device that doesn't line up is my pi4 with 32-bit OS reports `armv7l` but the binary is labeled `armhf`

I'm considering changing the wget command to
`sudo wget https://github.com/AppImage/AppImageKit/releases/download/continuous/appimagetool-$(uname -m).AppImage -O /usr/local/bin/appimagetool`

That would seem to get us 3/4 architectures correct

**Comment by starbasessd** [comment]
You could fix it in the build.sh (re-download it and re-set +x)? Or script a setup.sh with the pre-requisites?

**Comment by starbasessd** [comment]
I use apt-cacher-ng (it's part of my sources.list and other apt-defined files in all my linux setups), but isn't looked at by the hard-coded ubuntu downloaders... Any way to add it, either to the build.sh, or somewhere else?

**Comment by starbasessd** [comment]
Then apt-cacher-ng is a 'would be nice' or 'enhancement', or 'new feature' but I just did a quick "sed" line to fix it in build.sh for me.

**Comment by goldstar611** [comment]
> Then apt-cacher-ng is a 'would be nice' or 'enhancement', or 'new feature' but I just did a quick "sed" line to fix it in build.sh for me.

What change did you make in build.sh? Are you setting the `http_proxy` environment variable?

**Comment by starbasessd** [comment]
Nope, didn't see that, where is it?
I used 
>sed -i 's|http://ports|http://my_ip_address:3142/ports|g' ./build.sh

and
>sed -i 's|http://archive|http://my_ip_address:3142/archive|g' ./build.sh

Seems to work for me... If there is an easier way...

**Comment by starbasessd** [comment]
Not in current build.sh...

**Comment by goldstar611** [comment]
oh I see, you updated the urls in the build script. It's actually easier that that. Try this instead

```
http_proxy=http://localhost:3142 ./build.sh
```

I was able to build all 4 architectures offline using apt-cacher-ng

**Comment by starbasessd** [comment]
Another thing to add to the instructions!

**Comment by starbasessd** [comment]
Don't forget t at least note the different versions of appimagetool, as all 3 Debian based *nix's listed can run in all the architectures (aarch64, armhf/arm7, i686, and x86-64)
aarch64 on Pi4B, Pi3B & Pi3B+, and PiZeroW2, running 64 Bit OS (RPiOS Bullseye 64 bit, (to be tested on Ubuntu & Debian still))
armhf/arm7 on any Pi running RPiOS Buster Legacy and RPiOS Bullseye 32 bit
i686 on Intel / AMD running any Debian based 32 bit OS
x86-64 on Intel / AMD running any Debian based 64 bit OS

**Comment by goldstar611** [comment]
- Proxy notes added to the front page.
-I added the different architecture notes to the pre-built binaries block and moved that to the top. They are most relevant there (i.e. users who just want to help test rather than build)
- I updated the download of "appimagetool" to use `uname -m` and put an extra line to change armv7l to armhf.

I'm going to close this ticket out. If you think of anything else just open a new ticket! Thanks for everyone's help
